### PR TITLE
Mlgym#223 model card

### LINKED ISSUE #223: Data format specification of the modelcard for the endpoint
Let's have a call after the specification.

### PR DESCRIPTION
**Pull request by moinam**
Model Card API Implementation.
closes #223 

API:
_GET /model-card/<grid_search_id>/<experiment_id>_

Model card JSON structure:

{
    "model_details": {
        "model_description": < description of the model >,
        "model_version": < model version number >,
        "grid_search_id": < grid search id>,
        "train_date": < Training Date >,
        "source_repo": < Link to sorce code >,
        "train_params": < model trainable parameters >
    },
    "dataset_details": {
        "dataset_splits": {
            "split_config": < split configuration in dataset > ,
            "splits_percentage": < split percentage information >
        },
        "considered_dataset": < name of the dataset >,
        "label_distribution": < label distribution >
    },
    "training_details": {
        "hyperparams": < hyperparameters >,
        "loss_func": < name of loss function >,
        "optimizer": < name of the optimizer >
    },
    "eval_details": {
        "loss_funcs": < list of loss functions >,
        "metrics": [
            {
                "name": < metric name >,
                "params": < metric params >
            }
        ]
    },
    "experiment_environment": {
        "system_env": {
            "system_info": < Operating system > ,
            "architecture": < system architecture > ,
            "machine_type": < machine type Ex:AMD64 >,
            "processor": < processor name > ,
            "num_processor_cores": < Number of processor cores > ,
            "ram": < ram in GB > ,
            "python-version": < version > ,
            "python-packages": < List of pip packages > ,
            "CUDNN_version": < cudnn version > ,
            "num_cuda_device": < number of cuda devices > ,
            "cuda_device_list": [{
                "name": < name of GPU > ,
                "total_memory": < Total GPU memory in GB > ,
                "memory_allocated": < GPU memory allocated in GB > ,
                "memory_reserved": < GPU memory reserved in GB > ,
                "max_memory_reserved": < Max GPU memory reserved in GB >
            }]
        },
        "carbon_footprint": < Carbon footprint >,
        "entry_point_cmd": < Commands executed >
    }
}

### CODE REVIEW

File: example/grid_search_example/gs_config.yml [hunk 1]
```diff
@@ -5,6 +5,11 @@ global_config:
   model_prediction_key: &model_prediction_key_anchor model_prediction_key
   postprocessing_argmax_key: &postprocessing_argmax_key_anchor postprocessing_argmax_key
 
+model_info:  
+  model_description: "Testing 1 2 3 Hopefully it works."
+  model_version: 1.0.0
+  source_repo: "Github repo link Ex: https://github.com/mlgym/mlgym/tree/dev"
```
Comment (le1nux) [inline example/grid_search_example/gs_config.yml:11]: would be nice if we could retreive the source repo automatically in the future

File: src/ml_gym/util/util.py [hunk 2]
```diff
@@ -61,6 +106,154 @@ def create_system_info() -> Dict:
...
+        model_card["loss_func"] = self.loss_func
+        model_card["optimizer"] = self.optimizer
+        return model_card
+
+@dataclass
+class EvalDetails:
+    loss_funcs: list = None
+    metrics: list = None
+
+    def toJSON(self) -> Dict:
```
Comment (le1nux) [inline src/ml_gym/util/util.py:127]: the function is called toJSON but we are not returning a JSON string. We return a Dict. 
Do we really need this method? Could you check if calling `json_string = json.dumps(eval_details_object)` would work out of the box?

File: src/ml_gym/util/util.py [hunk 3]
```diff
@@ -61,6 +106,154 @@ def create_system_info() -> Dict:
...
+            """
+            Function to initialize ModelDetails object.
+            :params:
+                    grid_search_id (str): Grid Search ID created for the run.
+                    gs_config (dict): Grid Search configuration.
+            :returns:
+                    obj (ModelDetails): initialized model details object.
+            """
+            try:
+                if model != None:
```
Comment (le1nux) [inline src/ml_gym/util/util.py:174]: wrong way to check for None. 
it has to be: `if model is not None`

File: src/ml_gym/util/util.py [hunk 4]
```diff
@@ -61,6 +106,154 @@ def create_system_info() -> Dict:
...
+                    pytorch_total_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
+                else:
+                    pytorch_total_params = 0
+                train_date = datetime.now().strftime("%d-%m-%Y")
+                model_info = gs_config["model_info"]
+                return ModelDetails(model_description = model_info["model_description"], model_version = model_info["model_version"], grid_search_id = grid_search_id, train_date = train_date, source_repo = model_info["source_repo"], train_params= pytorch_total_params)
+            except Exception as e:
+                print("Error while fetching Model Details for Model card.", e)
```
Comment (le1nux) [inline src/ml_gym/util/util.py:182]: we should propagate the error and not "log" it as this low level in the call stack

Comment (le1nux) [reply src/ml_gym/util/util.py:182]: also a specific Exception would make more sense. What kind of things could go wrong here? 
As workaround we could catch for Exception and then reraise the Exception with a more specific exception type

File: src/ml_gym/util/util.py [hunk 5]
```diff
@@ -61,6 +106,154 @@ def create_system_info() -> Dict:
...
+                    obj (DatasetDetails): initialized dataset details object.
+            """
+            try:
+                dataset_splits = {"split_config": exp_config["dataset_iterators"]["config"]["split_configs"], 
+                              "splits_percentage": exp_config["splitted_dataset_iterators"]["config"]["split_configs"]}
+            
+                return DatasetDetails(considered_dataset = exp_config["dataset_iterators"]["config"]["dataset_identifier"], dataset_splits = dataset_splits)
+            except Exception as e:
+                print("Error while fetching Dataset Details for Model card.", e)
+                return DatasetDetails()
```
Comment (le1nux) [inline src/ml_gym/util/util.py:200]: see above

File: src/ml_gym/util/util.py [hunk 6]
```diff
@@ -61,6 +106,154 @@ def create_system_info() -> Dict:
...
+    def toJSON(self) -> Dict:
+        model_card = {}
+        model_card["model_details"] = self.model_details.toJSON()
+        model_card["dataset_details"] = self.dataset_details.toJSON()
+        model_card["training_details"] = self.training_details.toJSON()
+        model_card["eval_details"] = self.eval_details.toJSON()
+        model_card["experiment_environment"] = self.experiment_environment.toJSON()
+        return model_card
+
+class SystemEnv:
```
Comment (le1nux) [inline src/ml_gym/util/util.py:150]: Should we rename this? 
ModelCardFactory?

File: src/ml_gym/util/util.py [hunk 7]
```diff
@@ -61,6 +106,154 @@ def create_system_info() -> Dict:
...
+            """
+            Function to initialize TrainingDetails object.
+            :params:
+                    exp_config (dict): Experiment configuration.
+            :returns:
+                    obj (TrainingDetails): initialized training details object.
+            """
+            try:
+                hyperparams = {}
+                hyperparams["optimizer"] = exp_config["optimizer"]["config"]["params"]
```
Comment (le1nux) [inline src/ml_gym/util/util.py:212]: why did you hardcode this for optimizer only? There are various places where we might have a parameter sweep. E.g., different loss functions, different model paramters (e.g., activation functions), different seeds etc. 

So as I suggested in our previous meeting, we need to create this dynamically without hardcoding to certain components. We can get the information dynamically from the config of a single experiment (concrete value for a sweep) and the grid search yaml (contains all the sweeps and the ranges).

File: src/ml_gym/util/util.py [hunk 8]
```diff
@@ -61,6 +106,154 @@ def create_system_info() -> Dict:
...
+            try:
+                for loss_func_config in exp_config["eval_component"]["config"]["loss_funs_config"]:
+                    loss_funcs.append(loss_func_config["tag"])
+            
+                for metric_config in exp_config["eval_component"]["config"]["metrics_config"]:
+                    metrics.append({"name": metric_config["tag"], "params": metric_config["params"]})
+
+                return EvalDetails(loss_funcs = loss_funcs, metrics = metrics)
+            except Exception as e:
+                print("Error while fetching Eval Details for Model card.", e)
```
Comment (le1nux) [inline src/ml_gym/util/util.py:237]: we need improved exception handling. see above.

File: src/ml_gym/util/util.py [hunk 9]
```diff
@@ -61,6 +106,154 @@ def create_system_info() -> Dict:
...
+            dataset_details = update_dataset_details(exp_config = exp_config)
+            training_details = update_training_details(exp_config = exp_config, gs_config = gs_config)
+            eval_details = update_evaluation_details(exp_config = exp_config)
+            model_card = ModelCard(
+                model_details = model_details,
+                dataset_details = dataset_details,
+                experiment_environment = experiment_env,
+                training_details = training_details,
+                eval_details =eval_details
+            )
```
Comment (le1nux) [inline src/ml_gym/util/util.py:252]: I like how you separated this into different private functions :+1:  It improves the readability a lot

**Review by le1nux — CHANGES_REQUESTED** [review]
Really good work! We are getting there with the model cards. 
The only major thing that I found was that the hyperparameters are not inferred from the yaml config and the experiment config. In fact, currently they are hardcoded the way you implemented it. This should be improved. Other than that I only left some minor comments.

File: src/ml_gym/util/util.py (hunk 3 above)
Comment (moinam) [reply src/ml_gym/util/util.py:174]: ok updated it. Thank you.

File: src/ml_gym/util/util.py (hunk 6 above)
Comment (moinam) [reply src/ml_gym/util/util.py:150]: yes ofcourse

File: src/ml_gym/util/util.py (hunk 9 above)
Comment (moinam) [reply src/ml_gym/util/util.py:252]: thank you. :)

File: example/grid_search_example/gs_config.yml (hunk 1 above)
Comment (moinam) [reply example/grid_search_example/gs_config.yml:11]: Yes, I will try to give it another go and see if I ca get it dynamically from a python package if not I will keep it as it is and create an issue for further improvement.

File: src/ml_gym/util/util.py (hunk 7 above)
Comment (moinam) [reply src/ml_gym/util/util.py:212]: Well, this is not actually hard-coded but this was done to get the params of a particular sweep from the experiment config. Since we wanted to change the JSON structure I have done this to get something like this: 
"hyperparams": {
            "optimizer": {
                "lr": 1
            }

Comment (moinam) [reply src/ml_gym/util/util.py:212]: this information is being fetched dynamically but put into a static JSON structure for the model card. I did not compare it with the sweeps in grid search yaml, which I will do now.

File: src/ml_gym/util/util.py (hunk 8 above)
Comment (moinam) [reply src/ml_gym/util/util.py:237]: done

File: src/ml_gym/util/util.py (hunk 5 above)
Comment (moinam) [reply src/ml_gym/util/util.py:200]: done

File: src/ml_gym/util/util.py (hunk 4 above)
Comment (le1nux) [reply src/ml_gym/util/util.py:182]: yeah exception handling is good. In this case, I think that if there is an error in the model card retrieval, it is fine to let the execution crash with an exception, cause in that case something is really wrong within the training run and it would make sense to kill it. What kind of cases can we expect, that would cause an error here?

File: src/ml_gym/util/util.py (hunk 7 above)
Comment (le1nux) [reply src/ml_gym/util/util.py:212]: `exp_config["optimizer"]["config"]["params"]` the path is still harcoded. There might be configs that have a path like `exp_config["optimizer"]["config_2"]["paramsXY"]`.  We need to make sure that we get the path information from the grid search config

File: src/ml_gym/util/util.py (hunk 4 above)
Comment (moinam) [reply src/ml_gym/util/util.py:182]: ok, I will then make the changes accordingly. The only case where we can get an error is if the YAML and JSON file structure changes and the places where I am trying to retrieve the information are broken or not there.

File: src/ml_gym/util/util.py (hunk 7 above)
Comment (moinam) [reply src/ml_gym/util/util.py:212]: ok will try to  see how to fix this
